### PR TITLE
Fix lambda perftests for number of functions and versions

### PR DESCRIPTION
**Pull request by joe4dev**
## Motivation

The initial tests for exploring the maximum number of functions and function versions supported by LocalStack had some major flaws and lacked proper validation.
These tests help to test prebuilding strategies, explore LocalStack limits, and debug potential container cleanup issues.

## Changes

* Systematically test the number of functions and function versions for both sync and async invokes using appropriate validations
* Remove all unpredictable `time.sleep` and replace them with proper polling using an improved S3 integration validation
* Fix a missing modification in the number of function version test such that new versions were never created
* Improve [lambda_s3_integration.py](https://github.com/localstack/localstack/compare/improve-lambda-perftests?expand=1#diff-cf366852d272de8ff38bd74fbdba561e834bdd8563b12b524f2f3d8d182ae153)
  * More efficient using spooled file for testing file size upload with `file_size_bytes` (inspired by a customer scenario)
  * Persist the received event by default otherwise
* Validated all tests with small workloads against AWS: We want to ensure the functionality has parity wherever possible and feasible without running expensive workloads against AWS

## Testing

1. Enable performance testing using `TEST_PERFORMANCE=1`
2. Run these tests (potentially adjust workload size in first line):
* `test_number_of_function_versions_sync`
* `test_number_of_function_versions_async`
* `test_number_of_functions_sync`
* `test_number_of_functions_async`

These performance tests are not executed in the CI because they take too long for regular CI runs.

### CODE REVIEW

**Comment by joe4dev** [comment]
Ignoring the failing `coverage/coveralls` step because none of these tests are executed in CI

File: tests/aws/services/lambda_/test_lambda_performance.validation.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (dominikschubert) [inline tests/aws/services/lambda_/test_lambda_performance.validation.json:1]: nice :partying_face:

File: tests/aws/services/lambda_/test_lambda_performance.validation.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  "tests/aws/services/lambda_/test_lambda_performance.py::test_number_of_function_versions_async": {
+    "last_validated_date": "2024-02-08T17:18:44+00:00"
+  },
```
Comment (dominikschubert) [inline tests/aws/services/lambda_/test_lambda_performance.validation.json:4]: these tests are a case where having the test duration on AWS might be cool insight as well

**Review by dominikschubert — APPROVED** [review]
LGTM, nice work on validating these and cleaning them up a bit :+1:

File: tests/aws/services/lambda_/test_lambda_performance.validation.json (hunk 2 above)
Comment (joe4dev) [reply tests/aws/services/lambda_/test_lambda_performance.validation.json:4]: generally yes; the test duration here against AWS would be rather misleading though: 

```python
num_function_versions = 2 if is_aws_cloud() else 100
```

I think it is important to validate the functionality against AWS, but using a reduced workload saves us lots of 💸  if we ever regularly validate performance tests.

**Review by dfangl — APPROVED** [review]
Looking good!

**Pull request merged**